src/Components/Server/src/Circuits/CircuitIdFactory.cs

```diff
@@ -2,56 +2,114 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Security.Claims;
 using System.Security.Cryptography;
+using Microsoft.AspNetCore.Components.Server;
 using Microsoft.AspNetCore.DataProtection;
 using Microsoft.AspNetCore.WebUtilities;
-using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
     // This is a singleton instance
     // Generates strong cryptographic ids for circuits that are protected with authenticated encryption.
     internal class CircuitIdFactory
     {
+        private readonly IDataProtectionProvider _dataProtectorProvider = new EphemeralDataProtectionProvider();
+
         private const string CircuitIdProtectorPurpose = "Microsoft.AspNetCore.Components.Server";
 
         private readonly RandomNumberGenerator _generator = RandomNumberGenerator.Create();
         private readonly IDataProtector _protector;
 
-        public CircuitIdFactory(IDataProtectionProvider provider)
+        public CircuitIdFactory()
         {
-            _protector = provider.CreateProtector(CircuitIdProtectorPurpose);
+            _protector =  _dataProtectorProvider.CreateProtector(CircuitIdProtectorPurpose);
         }
 
         // Generates a circuit id that is produced from a strong cryptographic random number generator
         // we don't care about the underlying payload, other than its uniqueness and the fact that we
         // authenticate encrypt it using data protection.
         // For validation, the fact that we can unprotect the payload is guarantee enough.
-        public string CreateCircuitId()
+        public CircuitId CreateCircuitId()
         {
             var buffer = new byte[32];
             _generator.GetBytes(buffer);
             var payload = _protector.Protect(buffer);
 
-            return Base64UrlTextEncoder.Encode(payload);
+            return new CircuitId(buffer, GetHash(buffer), payload);
+        }
+
+        private static byte[] GetHash(byte[] buffer)
+        {
+            var hashAlgorithm = SHA256.Create();
+            var hash = hashAlgorithm.ComputeHash(buffer);
+            return hash;
+        }
+
+        public bool ValidateCircuitId(string circuitId, string cookie, out CircuitId id)
+        {
+            id = FromCookieValue(cookie);
+            var rawRequestId = Base64UrlTextEncoder.Decode(circuitId);
+            var requestTokenBytes = Base64UrlTextEncoder.Decode(id.RequestToken);
+            if (CryptographicOperations.FixedTimeEquals(rawRequestId, requestTokenBytes))
+            {
+                return true;
+            }
+            else
+            {
+                id = default;
+                return false;
+            }
         }
 
-        public bool ValidateCircuitId(string circuitId)
+        public bool ValidateCircuitId(string circuitId, ClaimsPrincipal user)
```

Comment:
Out of interest, have we got to the point of verifying that the current user's "name" claim matches the existing one on the circuit when reconnecting? Or is that a future enhancement?

Reply:
That's something we can do in the future, but it's not necessary. The way it works is we add an Identity to the request principal with a claim "bcid" which is the circuit id and we verify that the current principal has a claim that matches the provided circuit id.